testData.js

```diff
@@ -59,17 +63,26 @@ class TestData {
         if (testURL) {
             var response = await HTTP.getJSON(testURL);
             return new TestData(response, query);
-        }
-        else {
-            throw new Error("A 'url' parameter should be included on the query string.");
+        } else if (query.path) {
+            const fullTestPath = path.join(config.testsDir, query.path).normalize();
+            if (!(await exists(fullTestPath)) || !fullTestPath.startsWith(config.testsDir)) {

```

Comment:
what about sanitize like you did in `suite data`?
then you don't have to check if it starts with the test dir, right?
also, what does normalize do in the line above?

Reply:
Notice that query.path is a **path** that potentially can contain "/" in contrast to "tests" directory name in suiteData. Thus applying sanitize() will eliminate all "/" which is wrong behavior. normalize() resolve occurrences of .. and . to the directories this path traverse to (if possible). If after normalization fullTestPath starts in testsDir, the path is safe.